### PR TITLE
PKCS#12 Keystore Enhancement

### PR DESCRIPTION
**Pull request by AMHIT**
Added the feature to use PKCS#12 keystores as the repository for a client authentication connection

### CODE REVIEW

File: src/rdkafka_conf.c [hunk 1]
```diff
@@ -471,6 +471,16 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
 	  _RK(ssl.crl_location),
 	  "Path to CRL for verifying broker's certificate validity."
 	},
+	/* Andrea Minuto (PKCS12) */
```
Comment (edenhill) [inline src/rdkafka_conf.c]: Please remove your name from the comment, the git history already contains that information and it is not valuable in the source itself.

File: src/rdkafka_conf.c [hunk 2]
```diff
@@ -471,6 +471,16 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
...
+	{ _RK_GLOBAL, "ssl.pkcs12.keystore.location", _RK_C_STR,
```
Comment (edenhill) [inline src/rdkafka_conf.c]: We'll need to synchronize these property names with the Java client.
Ping @ijuma

File: src/rdkafka_transport.c [hunk 3]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
                 }
 	}
 
+	/* Andrea Minuto (PKCS12) */
+	if (rk->rk_conf.ssl.pkcs12.keystore_location) {
+		FILE *fp;
```
Comment (edenhill) [inline src/rdkafka_transport.c]: use 8 whitespaces for tab.

File: src/rdkafka_transport.c [hunk 4]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		EVP_PKEY *pkey;
+		X509 *cert;
+		STACK_OF(X509) *ca = NULL;
+		PKCS12 *p12;
+
+		rd_kafka_dbg(rk, SECURITY, "SSL",
+		"Loading client's keystore file from %s",
```
Comment (edenhill) [inline src/rdkafka_transport.c]: The indentation should match the vertical location of the open-paren

File: src/rdkafka_transport.c [hunk 5]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		rk->rk_conf.ssl.pkcs12.keystore_location);
+
+		if (!rk->rk_conf.ssl.pkcs12.keystore_password) {
+			rd_snprintf(errstr, errstr_size,
+				"The ssl.pkcs12.keystore.password is mandatory: ");
+			goto fail;
+		}
+
+		ERR_load_crypto_strings();
```
Comment (edenhill) [inline src/rdkafka_transport.c]: This should probably be done just once in rd_kafka_transport_ssl_init()

File: src/rdkafka_transport.c [hunk 6]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		if (!(fp = fopen(rk->rk_conf.ssl.pkcs12.keystore_location, "rb"))) {
+			rd_snprintf(errstr, errstr_size,
+				"ssl.pkcs12.keystore.location failed: ");
+			goto fail;
+		}
+
+		p12 = d2i_PKCS12_fp(fp, NULL);
+		fclose (fp);
```
Comment (edenhill) [inline src/rdkafka_transport.c]: remove space before open-paren

File: src/rdkafka_transport.c [hunk 7]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		if (!p12) {
+			rd_snprintf(errstr, errstr_size,
+				"Error reading PKCS#12 file: ");
+			goto fail;
+		}
+
+		if (!PKCS12_parse(p12, rk->rk_conf.ssl.pkcs12.keystore_password, &pkey, &cert, &ca)) {
+			rd_snprintf(errstr, errstr_size,
```
Comment (edenhill) [inline src/rdkafka_transport.c]: `p12` needs to be freed here.

File: src/rdkafka_transport.c [hunk 8]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		if (!rk->rk_conf.ssl.pkcs12.keystore_password) {
+			rd_snprintf(errstr, errstr_size,
+				"The ssl.pkcs12.keystore.password is mandatory: ");
+			goto fail;
+		}
+
+		ERR_load_crypto_strings();
+		if (!(fp = fopen(rk->rk_conf.ssl.pkcs12.keystore_location, "rb"))) {
+			rd_snprintf(errstr, errstr_size,
+				"ssl.pkcs12.keystore.location failed: ");
```
Comment (edenhill) [inline src/rdkafka_transport.c]: `"Failed to open ssl.pkcs12.keystore.location: %s: %s", ..key_store.location, rd_strerror(rd_errno());`

File: src/rdkafka_transport.c [hunk 9]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		EVP_PKEY *pkey;
+		X509 *cert;
+		STACK_OF(X509) *ca = NULL;
+		PKCS12 *p12;
+
+		rd_kafka_dbg(rk, SECURITY, "SSL",
+		"Loading client's keystore file from %s",
+		rk->rk_conf.ssl.pkcs12.keystore_location);
+
+		if (!rk->rk_conf.ssl.pkcs12.keystore_password) {
```
Comment (edenhill) [inline src/rdkafka_transport.c]: This should be enforced in rd_kafka_new()

File: src/rdkafka_transport.c [hunk 10]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		if (!PKCS12_parse(p12, rk->rk_conf.ssl.pkcs12.keystore_password, &pkey, &cert, &ca)) {
+			rd_snprintf(errstr, errstr_size,
+				"Error parsing PKCS#12 file: ");
+			goto fail;
+		}
+		PKCS12_free(p12);
+
+		r = SSL_CTX_use_certificate(ctx, cert);
+		if (r != 1) {
+			rd_snprintf(errstr, errstr_size,
```
Comment (edenhill) [inline src/rdkafka_transport.c]: Do we need to free the cert here?

File: src/rdkafka_transport.c [hunk 11]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+						"ssl.pkcs12.keystore.location failed: ");
```
Comment (edenhill) [inline src/rdkafka_transport.c]: `"Failed to use ssl.pkcs12.keystore.location certificate: "`

File: src/rdkafka_transport.c [hunk 12]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		fclose (fp);
+		if (!p12) {
+			rd_snprintf(errstr, errstr_size,
+				"Error reading PKCS#12 file: ");
+			goto fail;
+		}
+
+		if (!PKCS12_parse(p12, rk->rk_conf.ssl.pkcs12.keystore_password, &pkey, &cert, &ca)) {
+			rd_snprintf(errstr, errstr_size,
+				"Error parsing PKCS#12 file: ");
```
Comment (edenhill) [inline src/rdkafka_transport.c]: `"Failed to parse PKCS#12 file: %s: ", ..location`

File: src/rdkafka_transport.c [hunk 13]
```diff
@@ -920,6 +920,61 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+		r = SSL_CTX_use_certificate(ctx, cert);
+		if (r != 1) {
+			rd_snprintf(errstr, errstr_size,
+						"ssl.pkcs12.keystore.location failed: ");
+			goto fail;
+		}
+
+		r = SSL_CTX_use_PrivateKey(ctx, pkey);
+		if (r != 1) {
+			rd_snprintf(errstr, errstr_size,
```
Comment (edenhill) [inline src/rdkafka_transport.c]: `"Failed to use ssl.pkcs12.keystore.location private key: "`

File: src/rdkafka_transport_int.h [hunk 14]
```diff
@@ -35,6 +35,9 @@
 #if WITH_SSL
 #include <openssl/ssl.h>
 #include <openssl/err.h>
+/* Andrea Minuto (PKCS12) */
```
Comment (edenhill) [inline src/rdkafka_transport_int.h]: dito

**Comment by edenhill** [comment]
Make sure to run the tests before pushing PR, this breaks the 0004 test suite (see CI links on PR page)

**Comment by AMHIT** [comment]
I checked why the first test failed.
The only error was:
3: ### Test "0066_plugins" failed at /project/repo/checkout/tests/testcpp.h:67:test_FAIL(): ### 3: set(plugin.library.paths) failed: No such configuration property: "plugin.library.paths" 3: ### Test random seed was 91200605 ###
In my environment this test doesn’t exist and I’m sure that I never changed anything about the ‘plugin.library.paths’ config property.
Could you help me to understand how to fix the error?
Thank you very much
Andrea

**Comment by edenhill** [comment]
I think that test failure is unrelated to your change, try rebasing your commit on latest master and see if it is still occurring.

I will push off merging this (after all comments fixed) until after the upcoming maintenance release.

File: src/rdkafka_conf.c [hunk 15]
```diff
@@ -471,16 +471,25 @@ static const struct rd_kafka_property rd_kafka_properties[] = {
...
+	},
+	{ _RK_GLOBAL, "ssl.keystore.password", _RK_C_STR,
+	_RK(ssl.keystore_password),
+	"Client's keystore (PKCS#12) password."
+	},
 #endif /* WITH_SSL */
 
         /* Point user in the right direction if they try to apply
          * Java client SSL / JAAS properties. */
+		/*
```
Comment (edenhill) [inline src/rdkafka_conf.c]: Remove this code instead of commenting it out

File: src/rdkafka_transport.c [hunk 16]
```diff
@@ -920,6 +920,68 @@ int rd_kafka_transport_ssl_ctx_init (rd_kafka_t *rk,
...
+
+		PKCS12_free(p12);
+
+		if (ca != NULL)
+			sk_X509_pop_free(ca, X509_free);
+
+		r = SSL_CTX_use_certificate(ctx, cert);
+		X509_free(cert);
+		if (r != 1) {
+			rd_snprintf(errstr, errstr_size,
```
Comment (edenhill) [inline src/rdkafka_transport.c]: the pkey is leaked here

**Comment by edenhill** [comment]
Big thanks for this @AMHIT !

**Pull request merged**